### DISCUSSION TITLE
Quick Tag!

### DISCUSSION THREAD

**Discussion by beville** [comictagger/comictagger#674] [General]
Hey, the Quick Tag concept it quite cool!  

I was able to get matches with : "`Cory Doctorow's Futuristic Tales of the Here and Now #001 - Anda's Game (2007).cbz`"

I know the Metron dev was going to (or did already?) add hashes to his DB, but the server-side hamming distance matching is a step beyond what I thought was talked about.  

I'm curious how well that will scale for a truly large set (many 100's of thousands) of hashes to match against?  Is there some sort of special index building going on when new covers are added that can used optimize the searches?

**Comment by lordwelch** [comment]
This enough hashes?
```plain
Loaded hashes from "/perm/hashes.gz" msgpack hashes ahashes: 923269 dhashes: 926981 phashes: 926691
```

**Reply by beville** [reply]
That's a lot!  So I guess the matches aren't computationally expensive.  Very cool!

**Reply by lordwelch** [reply]
It uses 8 maps per type of hash to trim down the amount of direct hamming distance calculations it has to do.
For example if you take an 8 bit hash
```
1010 0110
```
and you need to find all hashes that are within a hamming distance of 1 you can cut it in half and find all the hashes that match the first or second halves exactly
Each of these match one half of the hash which guarantees that we find all hashes with a distance of 1 or less and then we can do a full check on the resulting hashed to trim down the results to what we want
```
1011 0110
1010 1110
1010 000
```

The server splits it 8 times
ram usage is at 1Gb but I think that's string duplication and the new version of go has a new library for helping with that

**Reply by lordwelch** [reply]
So I realized that just iterating through all 2.7 Million hashes (when all 3 hashes are sent) is still very quick so ram usage is now down to 197MB and it only takes 57.5ms to search through them. Using the maps to reduce the total number of lookups is still effective, it takes it down to ~10ms at 400MB ram, but at the moment I care more about ram usage.

**Comment by lordwelch** [comment]
> I know the Metron dev was going to (or did already?) add hashes to his DB, but the server-side hamming distance matching is a step beyond what I thought was talked about.

Yes he has the hashes available for all the comics on metron.cloud, it was done more to make it so that images don't need to be downloaded not to have robust matching with a hamming distance. I haven't gotten to revamping the search algorithm to take advantage of it yet.

**Comment by beville** [comment]
I've always wondered what the false positive rate would be with a system like this, where you're checking the visual hash against _all_ other comics, instead just a select few.  What have you observed so far?  Are the hashes pretty unique (meaning not (m)any other matches within a short distance)?

**Reply by lordwelch** [reply]
There aren't very many with an exact match, when testing a subset of my library of digital comics compared to ComicVine's covers for any of the hash types (ahash, dhash, phash).

However usually if it has a match it's within 8 and it's usually <=4 I'd guess about 40% of the time this works. Now sometimes there are a lot more that match within 8, if you compare between any of these matches to each other they will typically have a hamming distance of ~16 and this cover matches in-between them.

So if there is a single result that is <= 4 it assumes it's correct. Aggressive filtering checks to see if there are > 4 matches grouped by distance and then returns the closest group which may then return only a single <=4 match, this takes it up closer to 50%. For the rest of the matches ~90% have the correct match in the list but there are either other results or it's not close enough, so if it can match the series name and number from the comic it will automatically tag it.

I've found that some comics that I've had trouble finding because the name is different than on CV are found easily.

In general the ahash causes the most false positives and could probably be removed without reducing the number of correct matches by much, however I've found that the differences between covers are varied enough that it typically benefits from being able to use both the dhash and phash

**Reply by lordwelch** [reply]
CV also seems to be stricter on their rate limiting, they now hard-block at 200 requests an hour

**Reply by lordwelch** [reply]
I've since found some comics that only matched on the ahash so it might be better to request all the hashes and only fall back to the ahash if no other matches can be found
